packages/editor/src/components/post-locked-modal/index.js

```diff
@@ -44,18 +50,75 @@ export default function PostLockedModal() {
 			getEditedPostAttribute,
 			getEditorSettings,
 		} = select( editorStore );
-		const { getPostType } = select( coreStore );
+		const {
+			getPostType,
+			__experimentalIsCurrentUserSessionActive,
+		} = select( coreStore );
 		return {
+			isDraft: [ 'draft', 'auto-draft' ].includes(
+				getEditedPostAttribute( 'status' )
+			),
 			isLocked: isPostLocked(),
 			isTakeover: isPostLockTakeover(),
 			user: getPostLockUser(),
 			postId: getCurrentPostId(),
 			postLockUtils: getEditorSettings().postLockUtils,
 			activePostLock: getActivePostLock(),
 			postType: getPostType( getEditedPostAttribute( 'type' ) ),
+			isCurrentUserSessionActive: __experimentalIsCurrentUserSessionActive(),
 		};
 	} );
 
+	useEffect( () => {
+		const checkUserLoginHookName = `core/editor/check-user-login/post-locked-modal-${ instanceId }`;
+
+		/**
+		 * Binds to the Heartbeat Tick event and checks if the user is logged in or not.
+		 *
+		 * @param {Object} data Data received in the heartbeat request
+		 */
+		function checkUserLogin( data ) {
+			// Bail early, if the heartbeat has neither auth check data attribute nor nonce expiry attribute.
+			if (
+				! [ 'wp-auth-check', 'nonces_expired' ].some( ( attr ) =>
+					Object.prototype.hasOwnProperty.call( data, attr )
+				)
+			) {
+				return;
+			}
+
+			if ( ! data[ 'wp-auth-check' ] && isCurrentUserSessionActive ) {
+				// Update login state to false,
+				// if server does not respond with auth check
+				// and current login state is true.
+				updateCurrentUserSession( { isActive: false } );
+			} else if ( ! isCurrentUserSessionActive && data.nonces_expired ) {
+				// If current login state is false and nonce is renewed,
+				// trigger a heartbeat tick to check for latest post lock.
+				// Because first heartbeat will only renew the nonce.
+				// It will not return any errors if the post is already locked by other user or not.
+				if ( window.wp?.heartbeat?.connectNow ) {
+					triggerHeartbeatTick( true );
+					window.wp.heartbeat.connectNow();
+				}
+			} else if (
+				! isCurrentUserSessionActive &&
+				heartbeatTickTriggered &&
+				data[ 'wp-auth-check' ]
+			) {
+				// If current login state is false and server responds with auth check for previously triggered heartbeat,
+				// update login state to true.
+				updateCurrentUserSession( { isActive: true } );
+			}
+		}
```

Comment:
ℹ️  This whole function (which is managing the current user session in data store) can be questionable and there could be bette way to handle this. So I'm happy to take any advise of improvement here.